### DISCUSSION TITLE
Questions regarding NASA Data and Orion Pipelines

### DISCUSSION THREAD

**Discussion by github-user-29A** [sintel-dev/Orion#474] [Q&A] [answered]
Hello, 
 
I have a few questions regarding the Orion Library:
 - What is the sampling frequency of the NASA data supplied with the library?
 - Why are the timestamps for the NASA data as big as they are (is each measurement really six hours apart)?
 - Does Orion work with any time series data (both actual data values as well as the timestamps)? Does the data need to have a certain format besides having a timestamp and value column? (a specific sampling frequency for example)
 - During testing i input one of my signals (a recording of a ballistocardiographic signal; basically a low-frequency alternative to an electrocardiography) with VERY large artificial outliers (signal values are all <1, the outliers have a value of amplitude+2) into the ARIMA, AER andTADGAN pipelines. AER and TADGAN trained for 5 epochs, ARIMA predicted 1 step. 
   While AER and ARIMA found all anomalies, while TADGAN only found 3 out of 10. Is TADGAN not as good as the others?

Thank you in advance.

**Comment by sarahmish** [comment, marked as answer]
Hello @github-user-29A! Thank you for opening this discussion topic!

### NASA Data
We have the notebook used to convert and store the NASA data available [here](https://github.com/sintel-dev/Orion/blob/master/tutorials/Convert_NASA_Data_to_Orion_Format.ipynb).
* To my knowledge, the sampling frequency of the NASA is unknown. I would refer to the [original paper](https://arxiv.org/pdf/1802.04431.pdf) in case it does supply this information.
* In Orion, we synthetically added a `timestamp` column to the NASA data. The choice of six hours is arbitrary and does not exists in the originally distributed dataset by NASA.

### Data Preprocessing
* Additional information about the data need to be set in the hyperparameter field. For example, for sampling frequency, the `interval` hyperparameter in the primitive `mlstars.custom.timeseries_preprocessing.time_segments_aggregate#1` needs to be adjusted for correct pre-processing. Additional documentation can be found [here](https://sintel.dev/Orion/user_guides/primitives_pipelines/pipelines.html#pipelines-for-custom-data).
* Each pipeline differs from one another and the selection is based on the user's data. It is expected that not all pipelines will be able to find all the anomalies (or even the same subset of anomalies). According to our benchmark, `AER` currently surpasses other pipelines within some of the public datasets. For more details, see the [AER paper](https://arxiv.org/pdf/2212.13558.pdf) which presents a nice summary on the successes and limitations of the pipelines we have in this library in Section 4.

**Reply by github-user-29A** [reply]
Many thanks for your reply. 
Is is possible to run an entire data set through a pipeline?
I have a data set of 40 csv-files and so far I've only figured out how to train and predict on a single file (the detect anomalies and fit functions only take a data frame with time and values). I want to give the pipeline a few more files to train and predict on at once without having to concatinate them into one huge csv.

**Reply by sarahmish** [reply]
Are these csv files pertaining to the same signals? Or are they different? Here is how I would handle each situation:

**1. If the `csv` files are one signal split into 40 files**
In this case, I recommend concatenating these files into a single dataframe for cleaner training of the pipeline. If the `csv` files are large and you would like to process the data in batches, you can instantiate the Orion object once and call the `fit` function for each loaded `csv` file. It would be an iterative process as follows:
```python3
orion = Orion(..)

for file in files:
     df = pd.read_csv(file)
     orion.fit(df)
```

**2. If the `csv` files pertain to multiple signals**
Depending on whether these signals have the same characteristics as each other or not I would suggest different approaches. If the have the *same* characteristics, I would process them in batches similar to what I suggest in the previous point.
However, if their characteristics are different, meaning the `intervals` between one signal and another, or the range of values are different, then I would suggest creating a separate pipeline for each `csv` file.

I hope this addresses your question. Let me know if you have any follow-ups!

**Reply by github-user-29A** [reply]
Hello,

thank you for all your help so far.
In regards to the ARIMA model implemented in Orion. I am still trying to use it to find anomalies in ballistocardiographic data. While ARIMA found the artificial spikes it does not find naturally occuring anomalies within the signals.
Are there any strategies in regards to finding the right amount of parameters to use or is it just a matter of trial and error?

Here is a BCG signal as an example:
![BCG signal](https://github.com/sintel-dev/Orion/assets/44232013/e4a6c6cd-f394-4ef7-bca6-1671abcfe698)

Zoomed in Signal:
![BCG signal Zoomed](https://github.com/sintel-dev/Orion/assets/44232013/d1bf5f35-0f48-42e0-bd07-fe3df9aeb0da)

The data i use is from this link: https://ieee-dataport.org/open-access/bed-based-ballistocardiography-dataset
